sdk/tables/Azure.Data.Tables/src/TableEntity.cs

```diff
@@ -235,5 +236,56 @@ private static void EnforceType(Type requestedType, Type givenType)
                     $"Cannot return {requestedType} type for a {givenType} typed property."));
             }
         }
+
+        /// <summary>
+        /// Performs type coercion for numeric types.
+        /// Integers will be coerced to long or double.
+        /// Double and long values will accept integers values but remain double or long.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The type coercion is not allowed.</exception>
+        private static object CoerceType(object existingValue, object newValue)
+        {
+            if (!existingValue.GetType().IsAssignableFrom(newValue.GetType()))
+            {
+                object convertedValue = existingValue switch
+                {
+                    int _ => newValue switch
+                    {
+                        // All int values can be doubles
+                        double newDoubleValue =>  newDoubleValue,
+                        // All int values can be longs
+                        long newLongValue => newLongValue,
+                        _ => null
+                    },
+                    double _ => newValue switch
+                    {
+                        // if we already had a double value, preserve it as double even if newValue was an int.
+                        // example: entity["someDoubleValue"] = 5;
+                        int newIntValue => (double)newIntValue,
+                        _ => null
+                    },
+                    long _ => newValue switch
+                    {
+                        // if we already had a long value, preserve it as long even if newValue was an int.
+                        // example: entity["someLongValue"] = 5;
+                        int newIntValue => (long)newIntValue,
+                        _ => null
+                    },
+                    _ => null
+                };
+
+                if (convertedValue != null)
+                {
+                    return convertedValue;
+                }
+
+                // The value could not be safely converted.
+                throw new InvalidOperationException(string.Format(
+                    CultureInfo.InvariantCulture,
+                    $"Value ({existingValue.GetType()}) {existingValue} is not convertable to ({newValue.GetType()}) {newValue}."));
```

Comment:
I wonder if this message is helpful. As a customer assigning a value to a property I don't care if the old one is convertible to the new type because I'm overwriting the old value anyway. So why are you making me care about it?

Reply:
Is your suggestion to make it closer to the original message? 
"Value of type x cannot be assigned to property of type y"?

Reply:
I don't know :) Why do we have this limitation in the first place?

Reply:
The main reason is to prevent unintentional re-typing of properties. For example:
An entity is fetched from the service and a double typed property is updated to a whole number:
`entity["someLong"] = 5;`

The type of this property has now been silently changed to `int`. The thought is that if a property is ever a long or double, that the intention is to stay of that type because the service treats them differently. 

We could allow arbitrary assignment, but these restrictions make the type assignments more intentional. Prior to this change, we found in the UX study that the type enforcement was too restrictive, causing exceptions where we could have done the right thing instead.

Reply:
But you are adding support for silently changing the int to long. So why are some silent changes bad and the others good and what do I do if I really want to re-type the property?

Reply:
A change to `long` or `double` is explicit, whereas a change to `int` is not. For example:
```c#
// Intention of setting a double is clear here
entity["foo"] = 2.2;

// Intention of setting long is clear here
entity["foo"] = 5L;

// What did the user intend here?
entity["MyDouble"] = 4;
```

If you really want to retype an existing property, you could remove it first.
```c#
entity.Remove("MyDouble");

// this will be an int
entity["MyDouble"] = 4;
```

Reply:
Okay. What about something like "Unable to preserve the property type `{ODATA OR .NET TYPE HERE}` while assigning a value of type `{NEW VALUE}`. To discard the property type call Remove first."?

Reply:
BTW what about the following:
```
// Intention of setting DateTimeOffet is clear here
entity["foo"] = new DateTimeOffet();
```

Do we allow this?

Can you express the rule you are applying in one sentence?

Reply:
You raise a good point. We should allow any explicit type assignment. The essence of the rule I intended to enforce here is not to allow silent conversion of numeric types to `int`.

Reply:
Updated and totally eliminated the exception scenario.